src/hypothesis/extra/numpy.py

```diff
@@ -73,20 +96,206 @@ def is_scalar(spec):
+        if dtype == '?':
+            return st.just('?')
+        return st.builds(
+            dtype.format, st.sampled_from(endianness.replace('?', '<>')),
+            st.sampled_from(sorted(set(s // 8 for s in sizes))))
+
+    return (make_strat('?', bools) | make_strat('{}u{}', unsigned) |
+            make_strat('{}i{}', signed)).map(np.dtype)
+
+
+@st.defines_strategy
+def floating_dtypes(floats=True, complexs=True, real_sizes=(16, 32, 64)):
+    """Return a strategy for floating and complex dtypes.  Like
+    `integer_dtypes`, this strategy has a flag to disable each subtype.
+
+    real_sizes is the size in bits of the real part of a floating-point or
+    complex number.  A complex value will therefore be twice the size.
+
+    Complex halfs (a 16-bit real part) are not supported by numpy and will
+    not be generated by this strategy.
+    Larger floats (96 and 128 bit real parts) are not supported on all
+    platforms and therefore disabled by default.  To generate these dtypes,
+    include these values in the real_sizes argument.
+
+    """
+    assert real_sizes and all(s in (16, 32, 64, 96, 128) for s in real_sizes)
+    complex_sizes = [s * 2 for s in real_sizes if s != 16]
+    complexs &= bool(complex_sizes)
+
+    def make_strat(dtype, flag):
+        sizes = real_sizes if 'f' in dtype else complex_sizes
+        return st.sampled_from(sorted(set(s // 8 for s in sizes)))\
+            .map(dtype.format).map(np.dtype) if flag else st.nothing()
+
+    return make_strat('f{}', floats) | make_strat('c{}', complexs)
+
+
+@st.defines_strategy
+def timestamp_dtypes(max_period='Y', min_period='ns',
+                     allow_datetime=True, allow_timedelta=True):
+    """Return a strategy for datetime64 and timedelta64 dtypes, with various
+    precisions from year to attosecond."""
+    tr = TIME_RESOLUTIONS
+    selected = tr[tr.index(max_period):tr.index(min_period) + 1]
+    assert selected
+    assert allow_datetime or allow_timedelta
+
+    def make_strat(dtype, flag):
+        if not flag:
+            return st.nothing()
+        return st.sampled_from(selected).map(dtype.format).map(np.dtype)
+
+    return (make_strat('datetime64[{}]', allow_datetime) |
+            make_strat('timedelta64[{}]', allow_timedelta))
+
+
+@st.defines_strategy
+def string_dtypes(byte_strs=True, unicode_strs=True,
+                  min_len=0, max_len=16, endianness='?'):
+    """Return a strategy for generating string dtypes, both bytestrings and
+    unicode strings, of various lengths."""
+    assert endianness in ('?', '<', '=', '>')
+
+    def make_strat(dtype, flag):
+        return st.nothing() if not flag else st.builds(
+            dtype.format, st.sampled_from(endianness.replace('?', '<>')),
+            st.integers(min_len, max_len)).map(np.dtype)
+
+    return make_strat('{}S{}', byte_strs) | make_strat('{}U{}', unicode_strs)
+
+
+@st.defines_strategy
+def array_dtypes(subtype_strategy=scalar_dtypes(),
```

Comment:
I confess I don't really understand what's going on here, but that may be more unfamiliarity with numpy in general and this corner of it in particular. I'd find it useful if there were some tests explicitly for array\_dtypes that helped elaborate on what it was for/how to use it.